### ISSUE TITLE
Scrolling element into view for `Element Send Keys` is unclear

### ISSUE THREAD

**Issue by whimboo** [w3c/webdriver#1155]
https://w3c.github.io/webdriver/webdriver-spec.html#element-send-keys

```
14.3 Element Send Keys
------------------------
The Element Send Keys command scrolls into view the form control element and then sends the provided keys to the element.
```

The scroll attempt should not only be done for form elements, but at least also for `contenteditable` elements. Further there are focusable elements like HTMLAnchorElement, which we might also want to bring into view if possible.

So maybe we can make use of the following step?

> The Element Send Keys command scrolls into view the focusable element, and ...

Also it would be good to explain how to detect if an element is focusable. Simply by checking for the presence of the `focus` method on the element is not enough given that it is present even for HTMLElement.

**Comment by whimboo** [comment]
Just noticed that for `keyboard-interactable` we have a link to: https://html.spec.whatwg.org/#focusable-area. So maybe the following is more appropriate:

> The Element Send Keys command scrolls into view the `keyboard-interactable` element, and ...

**Comment by whimboo** [comment]
Further it would make sense to first check if the element is `keyboard-interactable` at all, before trying to scroll it into view. That way we would keep the position of the viewport, which we would lose if the order of steps are processed as currently layed out.

**Comment by andreastt** [comment]
This is another case of a confusing non-normative command introduction. The first paragraphs of the Element Send Keys command is not meant to be interpreted by implementors as normative. See https://github.com/w3c/webdriver/issues/1105 for a lot of similar confusion.

I’ve done some work in this area to remove confusing introductions. In the concrete case of Element Send Keys, if you look at the actual (normative) remote end steps it attempts to scroll all types of elements into view before interacting with it.

If you disregard the first paragraph under Element Send Keys, do the remote end steps make sense to you?

**Comment by whimboo** [comment]
Yes, it does. But there is only step 6 about this specific case. Only step 7 is a bit unclear to me given that I do not see a way how the keyboard interactability state will change by scrolling the element into view.

**Comment by andreastt** [comment]
The intention of step 7 is to re-run the [keyboard-interactability](https://w3c.github.io/webdriver/webdriver-spec.html#dfn-keyboard-interactable) check, like we do when finding elements, for an implementation-defined duration restricted upwards by the session implicit timeout.

Now, it is not entirely clear how to best _implement_ the keyboard-interactability concept because it depends on fairly complicated idea of focusable areas in HTML.

**Comment by whimboo** [comment]
> Now, it is not entirely clear how to best implement the keyboard-interactability concept because it depends on fairly complicated idea of focusable areas in HTML.

After a while of investigation and querying I found the following solution, which I got implemented for my patch on [bug 1414322](https://bugzilla.mozilla.org/show_bug.cgi?id=1414322):

1. An element is focusable (keyboard interactable) when it has the `tabIndex` property set to a value >-1
2. Or if the `isContentEditable` method on the element returns `true`
3. Or if it is the `body` element
4. Or if it is the document element

3 and 4 are defined by the spec, but 1 and 2 I had to figure out myself by reading the spec.

But maybe we should spin this out into a different issue now?

**Comment by andreastt** [comment]
Yes, I feel “what is a keyboard-interactable element” perhaps deserves its own discussion.

The four conditions you list are definitely some of the things that make an element keyboard-interactable. But if you see the table below https://html.spec.whatwg.org/#focusable-area there are certainly a whole range of other things it also affects.

**Comment by whimboo** [comment]
Here an update in how I solved the scroll into view and keyboard-interactability check in Firefox...

Given that the linked HTML specification for `focusable-area` has to be implemented by the browser, I was not eager to re-invent the wheel for our keyboard-interactability checks in Marionette. As such Marionette now simply uses `element.focus()` to determine that. It can be used because only focusable elements can get the focus, and as such become the `document.activeElement`. It means the check looks like that:

https://hg.mozilla.org/mozilla-central/diff/3cbe8afd86c7/testing/marionette/interaction.js#l1.25

Only `body` and `documentElement` have to be special-cased, which is as described in the webdriver spec.

A nice side-effect of using `focus()` here is that this call even tries to scroll the element into view by default. If it can't there is no reason to worry, because even if the element is outside of the viewport, it can receive focus, and key events.

https://html.spec.whatwg.org/multipage/interaction.html#dom-focus
